lambeq/training/nelder_mead_optimizer.py

```diff
@@ -0,0 +1,395 @@
+
+        The gradients are calculated with respect to the model
+        parameters.
+
+        Parameters
+        ----------
+        batch : tuple of Iterable and numpy.ndarray
+            Current batch. Contains an Iterable of diagrams in index 0,
+            and the targets in index 1.
+
+        Returns
+        -------
+        float
+            The calculated loss.
+
+        """
+
+        diagrams, targets = batch
+        diags_gen = flatten(diagrams)
+
+        # the symbolic parameters
+        parameters = self.model.symbols
+
+        relevant_params = set.union(
+            *[diag.free_symbols for diag in diags_gen]
+        )
+        mask = np.array(
+            [1 if sym in relevant_params else 0 for sym in parameters]
+        )
+
+        # Initialize ind, sim and fsim
+        if self.first_iter:
+            try:
+                for k in range(self.N + 1):
+                    self.fsim[k] = self.objective_func(diagrams,
+                                                       targets,
+                                                       self.sim[k])
+            except RuntimeError:
+                pass
+            finally:
+                self.ind = np.argsort(self.fsim)
+                self.sim = np.take(self.sim, self.ind, 0)
+                self.fsim = np.take(self.fsim, self.ind, 0)
+
+                self.nd = np.argsort(self.fsim)
+                self.fsim = np.take(self.fsim, self.ind, 0)
+
+                # sort so sim[0,:] has the lowest function value
+                self.sim = np.take(self.sim, self.ind, 0)
+                self.first_iter = False  # set flag to false
+
+        try:
+            if (
+                np.max(np.ravel(np.abs(self.sim[1:] - self.sim[0])))
+                <= self.xatol
+                and np.max(np.abs(self.fsim[0] - self.fsim[1:])) <= self.fatol
+            ):
+                raise ValueError(
+                    'xtol and ftol termination conditions are satisfied.'
+                )
+
+            xbar = np.add.reduce(self.sim[:-1], 0) / self.N
+            xr = (1 + self.rho) * xbar - self.rho * self.sim[-1]
+            xr = self.project(xr)
+            fxr = self.objective_func(diagrams, targets, xr)
+            doshrink = 0
+
+            if fxr < self.fsim[0]:
+                xe = (
+                    1 + self.rho * self.chi
+                ) * xbar - self.rho * self.chi * self.sim[-1]
+                xe = self.project(xe)
+                fxe = self.objective_func(diagrams, targets, xe)
+                if fxe < fxr:
+                    self.sim[-1] = xe
+                    self.fsim[-1] = fxe
+                else:
+                    self.sim[-1] = xr
+                    self.fsim[-1] = fxr
+            else:  # fsim[0] <= fxr
+                if fxr < self.fsim[-2]:
+                    self.sim[-1] = xr
+                    self.fsim[-1] = fxr
+                else:  # fxr >= fsim[-2]
```

Comment:
Can we do `elif`s here instead of nesting?

Reply:
Are you talking about line 304? I think we can but I also think the way it is written currently improves the readability of the logical flow. The else block handles the contraction.